Часть 1

In [4]:
#!pip install yargy
import pandas as pd
pd.set_option('display.max_colwidth', -1)
from yargy import Parser, rule, or_
from yargy.tokenizer import MorphTokenizer
from yargy.pipelines import morph_pipeline, caseless_pipeline
from yargy.interpretation import fact

     |████████████████████████████████| 51kB 4.2MB/s 
     |████████████████████████████████| 51kB 5.6MB/s 
     |████████████████████████████████| 7.1MB 7.8MB/s 


In [5]:
!wget https://raw.githubusercontent.com/mannefedov/compling_nlp_hse_course/master/data/pristavki.csv
data = pd.read_csv('pristavki.csv', header=None, names=['text'])

--2020-01-20 23:11:12--  https://raw.githubusercontent.com/mannefedov/compling_nlp_hse_course/master/data/pristavki.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2939556 (2.8M) [text/plain]
Saving to: ‘pristavki.csv’

pristavki.csv       100%[===================>]   2.80M  --.-KB/s    in 0.06s   

2020-01-20 23:11:13 (46.8 MB/s) - ‘pristavki.csv’ saved [2939556/2939556]



In [0]:

game = fact('game', ['name', 'release'])
releases = {'san andreas':'San Andreas', 'sanandreas':'San Andreas',
            'vice city': 'Vice City','vicecity': 'Vice City',
            '5': 'V', 'v': 'V', '4': 'IV', 'iv': 'IV',
            'shift': 'Shift', 'underground': 'Underground', 'most wanted': 'Most Wanted', 'carbon': 'Carbon', 'rivals':'Rivals'}
fifa = rule(morph_pipeline(['fifa', 'фифа', 'фифв', 'фифу']).interpretation(game.name),
            morph_pipeline(['10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', 'STREET']).interpretation(game.release))
NFS = rule(morph_pipeline(['NFS', 'Need for Speed', 'нфс']).interpretation(game.name),
           caseless_pipeline(releases).interpretation(game.release.normalized().custom(releases.get)))
GTA = rule(morph_pipeline(['gta', 'grand theft auto', 'гта']).interpretation(game.name),
           caseless_pipeline(releases).interpretation(game.release.normalized().custom(releases.get)))
GAME = or_(fifa, NFS, GTA).interpretation(game)
parser = Parser(GAME)

In [7]:
matches = []
list(parser.findall('GTA sanandreas'))[0].fact


game(name='GTA',
     release='San Andreas')

In [8]:
for sent in data.text:
    for match in parser.findall(sent):
        matches.append(match.fact)
for m in matches:
    print(m.name, m.release)

фифа 15
гта V
Фифв 13
гта IV
FIFA 16
GTA V
Need For Speed Shift
Fifa 11
Fifa 15
gta V
GTA V
GTA V
GTA V
GTA V
GTA V
гта V
GTA V
FIFA 16
FIFA 13
FIFA 16
Fifa STREET
GTA V
FIFA 16
FIFA 15
GTA V
Need For Speed Most Wanted
GTA IV
FIFA 17
gta V
Fifa 15
FIFA 16
FIFA 15
Need for Speed Rivals
GTA V
ГТА V
gta V
fifa 16
fifa 17
GTA V
GTA V
fifa 17
fifa 17
гта V
gta V
fifa 14
FIFA 14
FIFA 15
FIFA 12
GTA V
Fifa 14
Fifa 16
fifa 15
GTA V
FIFA 15
GTA V
FIFA 15
гта V
fifa 15
Gta V
NFS Rivals
Grand Theft Auto V
GTA V
GTA V
Grand Theft Auto V
GTA V
Фифа 15
FIFA 17
GTA V
GTA V
Fifa Street
gta V
fifa 15
nfs Rivals
GTA V
NFS Carbon
гта V
фифа 17
FIFA 15
GTA V
gta V
FIFA 16
Need for Speed Rivals
GTA V
Fifa 13
NFS Most Wanted
FIFA 16
FIFA 17
GTA V
GTA V
FIFA 14
gta Vice City
фифа 17
фифа 15
гта V
NFS Rivals
Gta V
fifa 16
FIFA 15
Grand Theft Auto V
Fifa 15
ГТА V
NFS Rivals
Need for speed Rivals
NFS Shift
GTA V
gta V
FiFa 13
GTA V
Need for Speed Most Wanted
GTA IV
Need for speed Most Wanted
FIFA 15
гта V
Fifa 

In [10]:
print(len(matches)) # количество найденных сущностей

1088


In [0]:
# То же самое, но вместе с ".optional" там, где идет информация о части (релизе), что делает ее наличие необязательным
game = fact('game', ['name', 'release'])
releases = {'san andreas':'San Andreas', 'sanandreas':'San Andreas',
            'vice city': 'Vice City','vicecity': 'Vice City',
            '5': 'V', 'v': 'V', '4': 'IV', 'iv': 'IV',
            'shift': 'Shift', 'underground': 'Underground', 'most wanted': 'Most Wanted', 'carbon': 'Carbon', 'rivals':'Rivals'}
fifa = rule(morph_pipeline(['fifa', 'фифа', 'фифв', 'фифу']).interpretation(game.name),
            morph_pipeline(['10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', 'STREET']).interpretation(game.release).optional())
NFS = rule(morph_pipeline(['NFS', 'Need for Speed', 'нфс']).interpretation(game.name),
           caseless_pipeline(releases).interpretation(game.release.normalized().custom(releases.get)).optional())
GTA = rule(morph_pipeline(['gta', 'grand theft auto', 'гта']).interpretation(game.name),
           caseless_pipeline(releases).interpretation(game.release.normalized().custom(releases.get)).optional())
GAME = or_(fifa, NFS, GTA).interpretation(game)
parser = Parser(GAME)

In [18]:
matches = []
list(parser.findall('GTA sanandreas'))[0].fact


game(name='GTA',
     release='San Andreas')

In [19]:
for sent in data.text:
    for match in parser.findall(sent):
        matches.append(match.fact)
for m in matches:
    print(m.name, m.release)

фифа 15
гта V
Фифв 13
гта IV
FIFA 16
GTA V
FIFA None
Need For Speed Shift
Fifa 11
Fifa 15
gta V
GTA V
Need For Speed None
Fifa None
GTA V
nfs None
Nfs None
GTA V
GTA V
GTA V
Fifa None
гта V
GTA V
FIFA 16
FIFA 13
FIFA 16
Fifa STREET
GTA V
GTA None
FIFA 16
FIFA 15
GTA V
Need For Speed Most Wanted
GTA None
Need For Speed None
Fifa None
GTA IV
FIFA 17
gta V
Fifa 15
FIFA 16
FIFA 15
Need for Speed Rivals
GTA V
Need for Speed None
ГТА V
gta V
fifa 16
fifa 17
GTA V
GTA V
fifa 17
fifa 17
гта V
NFS None
gta V
fifa 14
gta None
FIFA 14
FIFA 15
Fifa None
FIFA 12
GTA V
Fifa 14
Fifa 16
fifa 15
GTA V
FIFA 15
Fifa None
GTA V
FIFA 15
гта V
fifa 15
Gta V
NFS Rivals
Grand Theft Auto V
GTA V
GTA V
Grand Theft Auto None
Grand Theft Auto V
Grand Theft Auto None
NFS None
GTA V
gta None
nfs None
Фифа 15
FIFA 17
Gta None
GTA V
GTA V
NFS None
GTA None
Fifa Street
Fifa None
gta V
fifa 15
nfs Rivals
GTA V
NFS Carbon
гта V
фифа 17
FIFA 15
GTA V
gta V
FIFA 16
Need for Speed Rivals
ГТА None
GTA V
Nfs None
Fifa 13
NFS

In [20]:
print(len(matches))

1449


Можно сделать вывод, что правила для релизов игр в первом случае были прописаны неидеально (Охват 1088 из 1449). В те 1449 штук также могли входить игры без упоминания их подвидов (в объявлениях вроде "продаю приставку с играми NFS, Fifa, GTA"), но думаю, что таких мало, и надо просто смотреть какие еще виды игр и способы их написания существуют. Ещё наверняка были охвачены не все возможные опечатки наподобие "фифв" для "Fifa". 